lib/backend/sqlbk/background.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2018-2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sqlbk
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/trace"
+)
+
+// start background goroutine to track expired leases, emit events, and purge records.
+func (b *Backend) start(ctx context.Context) error {
+	tx := b.db.ReadOnly(ctx)
+	lastEventID := tx.GetLastEventID()
+	if tx.Commit() != nil {
+		return trace.ConnectionProblem(tx.Err(), "failed to query for last event ID")
+	}
+	b.buf.SetInit()
+	go b.run(lastEventID)
+	return nil
+}
+
+// run background process.
+// - Poll the database to delete expired leases and emit events every PollStreamPeriod (1s).
+// - Purge expired backend items and emitted events every PurgePeriod (10s).
+func (b *Backend) run(eventID int64) {
+	pollTicker := time.NewTicker(b.PollStreamPeriod)
+	defer pollTicker.Stop()
+
+	purgeTicker := time.NewTicker(b.PurgePeriod)
+	defer purgeTicker.Stop()
+
+	var err error
+	var loggedError bool // don't spam logs
+	for {
+		select {
+		case <-b.closeCtx.Done():
+			return
+		case <-pollTicker.C:
+			eventID, err = b.poll(eventID)
```

Comment:
Relying on an event ID sequence without gaps is not possible in most SQL databases. By default, auto-incrementing sequences in CockroachDB are not sequential. They are a 64-bit integer from the current timestamp and ID of the node executing the operation. CockroachDB, PostgreSQL, and MySQL do have a sequential ID mode, but it is not guaranteed to be sequential because each insert increases the sequence by one, even when the insert is not committed. This means that auto-incrementing types may leave gaps in a sequence.

I can think of two scenarios where an auth server could miss events. First, if two auth servers use different settings for PurgePeriod, PollStreamPeriod, or BufferSize (used for setting a limit on the number of events per poll period). Second, if there are more events to emit than the settings are capable of before a purge event.

For the first scenario, we could remove the ability to set PurgePeriod, PollStreamPeriod, and BufferSize via the YAML configuration. These types of settings are typically not documented, so maybe they could stay. But worst case we could change the JSON tag to not set a value from YAML so the variables always use the default. One way of another, each auth server should use the same settings.

For the second scenario, assuming the auth servers are using the same settings, we could detect whether the server is falling behind emitting events and reset. E.g. reset when this calculation is true because we know we’ll miss events:
`<number of events remaining to emit> / <limit per poll period> * PollStreamPeriod > PurgePeriod`

Reply:
Oof.  That's unfortunate.  Thanks for linking that doc.  This tidbit was especially surprising to me:

> If two transactions occur concurrently, CockroachDB cannot guarantee monotonically increasing IDs (i.e., first commit is smaller than second commit).

That seems crazy to me. Skipping is one thing, but permitting completely out of order insertions seems nuts.

With my limited understanding of SQL, I'm not sure what the full implications of this are... does this mean that we could have event IDs show up in the wrong order *for a given key*, or would any concurrent modification of the same key result in an `ErrRetry`?  E.g. if one auth server runs `put "k" => "v1"` and the other runs `put "k" => "v2"`... if the final value at `"k"` ends up being `"v2"` is it possible that the event ordering could result in the `put "k" => "v2"` event being ordered *before* the `put "k" => "v1"` event?  Or are two concurrent put operations still going to have one of them see an `ErrRetry`?

If it is possible for event sequencing *for a given key* to be wrong, that's a very serious concern for us, and we'll need to fix that in some way, even if it means sacrificing performance.  Fast is good, but teleport is security software, and that means that correctness has to always come before efficiency.

Reply:
IIUC, two concurrent transactions operating on the same key will result in an ErrRetry for one of them (the concurrent backend test suite function proves this works). So the scenario in the documentation when applied to teleport would be that two concurrent transactions operating on _different_ keys may order events incorrectly when they occur within 10 microseconds of each other.

The scenario described applies to the SERIAL data type. We could move to use a SEQUENCE instead for CockroachDB (I'd have to move some things around because the schemas for cockroach and postgres would be different). Although a sequence can still have gaps and ordering problems for different keys because the sequence is incremented atomically inside the transaction and not rolled back. We'd still have the same issue: keyA may increment before keyB but keyB commits before keyA. So I'm not sure what the advantage of a SEQUENCE is over SERIAL other than the values being closer together.

It sounds like we can keep things as they are since the issue only occurs with different keys.

I implemented the latency detection calculation described above with test coverage.